### PR TITLE
First-class Keras support

### PR DESCRIPTION
**Pull request by alsrgv**
Introduce package horovod.keras which provides functional parity between Keras and TensorFlow.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -146,9 +146,7 @@ Check your MPI documentation for arguments to the `mpirun` command on your syste
 
 Horovod supports Keras and regular TensorFlow in similar ways.
 
-**Note**: You must use `keras.optimizers.TFOptimizer` instead of native Keras optimizers.
-
-See a full training example [here](examples/keras_mnist.py).
+See a full training [simple](examples/keras_mnist.py) and [advanced](examples/keras_mnist_advanced.py) examples.
```
Comment (sblotner) [inline README.md]: remove "a"

File: horovod/keras/__init__.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        optimizer: Optimizer to use for computing gradients and applying updates.
+        name: Optional name prefix for the operations created when applying
+              gradients. Defaults to "Distributed" followed by the provided
+              optimizer type.
+        device_dense: Device to be used for dense tensors. Uses GPU by default
+                      if Horovod was build with HOROVOD_GPU_ALLREDUCE.
+        device_sparse: Device to be used for sparse tensors. Uses GPU by default
+                       if Horovod was build with HOROVOD_GPU_ALLGATHER.
+    """
+    # We dynamically create a new class that inherit from the optimizer that was passed in.
```
Comment (sblotner) [inline horovod/keras/__init__.py]: inherit --> inherits

File: horovod/keras/__init__.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    bcast_op = hvd.broadcast_global_variables(root_rank)
+    return K.get_session().run(bcast_op)
+
+
+def allreduce(value, name=None, average=True):
+    """
+    Perform an allreduce on a tensor-compatible value.
+
+    Arguments:
+        value: a tensor-compatible value to reduce.
```
Comment (sblotner) [inline horovod/keras/__init__.py]: a --> A

File: horovod/keras/__init__.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+def allgather(value, name=None):
+    """
+    Perform an allgather on a tensor-compatible value.
+
+    The concatenation is done on the first dimension, so the input values on the
+    different processes must have the same rank and shape, except for the first
+    dimension, which is allowed to be different.
+
+    Arguments:
+        value: a tensor-compatible value to gather.
```
Comment (sblotner) [inline horovod/keras/__init__.py]: a --> A

File: horovod/keras/__init__.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        name: Optional name prefix for the constants created by this operation.
+    """
+    allgather_op = hvd.allgather(tf.constant(value, name=name))
+    return K.get_session().run(allgather_op)
+
+
+def broadcast(value, root_rank, name=None):
+    """
+    Perform an broadcast on a tensor-compatible value.
```
Comment (sblotner) [inline horovod/keras/__init__.py]: an --> a

File: horovod/keras/__init__.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+    Arguments:
+        value: a tensor-compatible value to reduce.
```
Comment (sblotner) [inline horovod/keras/__init__.py]: a --> A

File: horovod/keras/__init__.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+               The shape of the input must be identical across all ranks.
+        root_rank: rank of the process from which global variables will be
```
Comment (sblotner) [inline horovod/keras/__init__.py]: rank --> Rank

File: horovod/keras/callbacks.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+            K.get_session().run(bcast_op)
+
+
+class MetricAverageCallback(keras.callbacks.Callback):
+    """
+    Keras Callback that will average metrics across all processes at the
+    end of the epoch. Useful in conjuction with ReduceLROnPlateau,
+    TensorBoard and other metrics-based callbacks.
+
+    Note: this callback must be added to the callback list before the
```
Comment (sblotner) [inline horovod/keras/callbacks.py]: this --> This

**Comment by alsrgv** [comment]
+@fchollet would appreciate a feedback if you have any. See usage examples [here](https://github.com/uber/horovod/blob/keras/examples/keras_mnist.py) and [here](https://github.com/uber/horovod/blob/keras/examples/keras_mnist_advanced.py)

**Comment by fchollet** [comment]
Will try to check it out tomorrow. Thanks for the heads up!

File: examples/keras_mnist.py [hunk 9]
```diff
@@ -64,19 +64,19 @@
...
-              optimizer=keras.optimizers.TFOptimizer(opt),
+              optimizer=opt,
               metrics=['accuracy'])
 
-# Broadcast variables from rank 0 to all other processes.
-K.get_session().run(hvd.broadcast_global_variables(0))
-
-callbacks = []
+callbacks = [
+    # Broadcast variables from rank 0 to all other processes.
```
[the remark this answers is not in the document]
Comment (alsrgv) [reply examples/keras_mnist.py]: Thanks for the feedback!  Added clarifications about the broadcast callback.

**Pull request merged**